### PR TITLE
Implement support for InheritDataTypeFromItemsAttribute

### LINKED ISSUE #10042: Unable to bind DataGridTextColumn
**Describe the bug**
Trying to bind DataGridTextColumn to a property on my class causes the following error:
`Import.axaml(16, 48): [XAMLIL] Unable to resolve property or method of name 'OtherAccountNumber' on type 'FinanceTracker.ViewModels.ImportModel'. Line 16, position 48.`

Doing the same with a textbox inside a DataGridTemplateColumn works without issue.

**To Reproduce**
``` csharp
public class ImportModel {
    public ObservableCollection<TransactionRecord> TransactionRecords { get; }
    
    public ImportModel () {
        TransactionRecords = new ObservableCollection<TransactionRecord> {
            new TransactionRecord() {OtherAccountNumber = "Testnumber"}
        };
    }
}

public class TransactionRecord {
    public string OtherAccountNumber { get; set; }
}
```

``` xaml
<Design.DataContext>
    <viewModels:ImportModel />
</Design.DataContext>
<DataGrid Items="{Binding TransactionRecords}" AutoGenerateColumns="True">
    <DataGrid.Columns>
        <DataGridTextColumn Header="Hello" Binding="{Binding OtherAccountNumber}">
        </DataGridTextColumn>
        <DataGridTemplateColumn Header="Test">
            <DataGridTemplateColumn.CellTemplate>
                <DataTemplate DataType="import:TransactionRecord">
                    <TextBlock Text="{Binding OtherAccountNumber}" />
                </DataTemplate>
            </DataGridTemplateColumn.CellTemplate>
        </DataGridTemplateColumn>
    </DataGrid.Columns>
</DataGrid>
```

- OS: Windows
- Version 11.0.0-preview4

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

Adds new API with compiler support:
```csharp
namespace Avalonia.Metadata;

/// <summary>
/// Instructs the compiler to resolve the compiled bindings data type for the item-specific properties of collection-like controls. 
/// </summary>
/// <remarks>
/// A typical usage example is a ListBox control, where <see cref="InheritDataTypeFromItemsAttribute"/> is defined on the ItemTemplate property,
/// allowing the template to inherit the data type from the Items collection binding. 
/// </remarks>
[AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
public sealed class InheritDataTypeFromItemsAttribute : Attribute
{
    /// <summary>
    /// Initializes a new instance of the <see cref="InheritDataTypeFromItemsAttribute"/> class.
    /// </summary>
    /// <param name="ancestorItemsProperty">The name of the property whose item type should be used on the target property.</param>
    public InheritDataTypeFromItemsAttribute(string ancestorItemsProperty)
    {
        AncestorItemsProperty = ancestorItemsProperty;
    }
    
    /// <summary>
    /// The name of the property whose item type should be used on the target property.
    /// </summary>
    public string AncestorItemsProperty { get; }
    
    /// <summary>
    /// The ancestor type to be used in a lookup for the <see cref="AncestorProperty"/>.
    /// If null, the declaring type of the target property is used.
    /// </summary>
    public Type? AncestorType { get; set; }
}
```

## What is the current behavior?

Currently ItemsControl and ItemsRepeater are hardcoded in the compiler.

## What is the updated/expected behavior with this PR?

Flexible support for custom collection types.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #10042
Helps with #10112

### CODE REVIEW

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs [hunk 1]
```diff
@@ -37,6 +37,10 @@ public static IXamlType UpdateCompiledBindingExtension(AstTransformationContext
                 bindingResultType = transformed.BindingResultType;
                 binding.Arguments[0] = transformed;
             }
+            if (binding.Arguments.Count > 0 && binding.Arguments[0] is XamlIlBindingPathNode alreadyTransformed)
```
Comment (kekekeks) [inline src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs:40]: Should probably be `else if`

Comment (maxkatz6) [reply src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs:40]: Shouldn't be a problem here, but yes, better to fix

Comment (maxkatz6) [inline src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/XamlIlBindingPathHelper.cs:40]:
```suggestion
            else if (binding.Arguments.Count > 0 && binding.Arguments[0] is XamlIlBindingPathNode alreadyTransformed)
```

**Review by kekekeks — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029234-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**